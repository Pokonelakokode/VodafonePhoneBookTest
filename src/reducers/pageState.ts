import {IPageStateActions, PageStateActionTypes} from "../actions/pageState";
import {getHash} from "../utils";

export type PageState = {
    selectedUser: number | null,
    search: string,
    letterFilter: string
}

const initialState:PageState = {
    selectedUser: parseInt(getHash().user) || null,
    search: "",
    letterFilter: ""
};


export default function pageState(state = initialState, action: IPageStateActions) {
    switch (action.type) {
        case PageStateActionTypes.SET:
            return {...state,[action.key]:action.data};
        default:
            return state
    }
}

