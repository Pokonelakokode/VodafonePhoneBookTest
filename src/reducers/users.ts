import {IUserActions, UserActionTypes} from "../actions/users";
import { User } from "../types";
import {loadFromStorage} from "../utils";

export default function users(state = loadFromStorage(), action: IUserActions):User[] {
    switch (action.type) {
        case UserActionTypes.LOAD:
            return action.users;
        case UserActionTypes.SET:
            return state.map(user => user.id === action.user.id ? action.user : user);
        case UserActionTypes.ADD:
            return [...state,action.user];
        case UserActionTypes.DELETE:
            return state.filter(user => user.id !== action.id);
        default:
            return state
    }
}