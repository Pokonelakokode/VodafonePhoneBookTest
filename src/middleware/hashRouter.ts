import {Dispatch, Middleware} from "redux";
import {RootActions} from "../actions";
import {PageStateActionTypes} from "../actions/pageState";
import {setHash} from "../utils";

export const hashRouter:Middleware = () => (next:Dispatch) => (action:RootActions) => {
    next(action);
    if(action.type === PageStateActionTypes.SET && action.key === "selectedUser"){
        action.data ? setHash({user:action.data}) : setHash();
    }
};