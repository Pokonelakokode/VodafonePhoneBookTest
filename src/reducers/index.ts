import { combineReducers } from "redux";
import pageState from "./pageState";
import users from "./users";

export const rootReducer = combineReducers({
    pageState,
    users
});

export type RootState = ReturnType<typeof rootReducer>