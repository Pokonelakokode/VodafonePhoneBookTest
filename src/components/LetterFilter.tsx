import * as React from "react";
import Letter from "./Letter";
import {useDispatch, useSelector} from "react-redux";
import { RootState } from "../reducers";
import {setPageState} from "../actions/pageState";

const LetterFilter:React.FC = () => {
    const {users,pageState:{letterFilter}} = useSelector((state:RootState) => state);
    const dispatch = useDispatch();
    const selectLetter = (letter:string) => dispatch(setPageState("letterFilter", letter));
    const letters = users
        .reduce((letters: string[], user) => {
            if (!letters.includes(user.name.charAt(0))) letters.push(user.name.charAt(0));
            return letters
        }, [])
        .sort((a, b) => a.localeCompare(b));
    return (
        <div className="col-12 letters">
            {letters.map(letter => (
                <Letter
                    key={letter}
                    selected={letterFilter === letter}
                    letter={letter.toUpperCase()}
                    selectLetter={selectLetter}
                />
            ))}
        </div>
    )
};

export default LetterFilter