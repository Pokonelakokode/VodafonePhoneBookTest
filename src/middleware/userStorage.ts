import {Dispatch, Middleware, MiddlewareAPI} from "redux";
import {RootActions} from "../actions";
import {UserActionTypes} from "../actions/users";
import {saveToStorage} from "../utils";
import {RootState} from "../reducers";

export const userStorage:Middleware = (store: MiddlewareAPI<any,RootState>) => (next:Dispatch) => (action:RootActions) => {
    const {DELETE, ADD, SET} = UserActionTypes;
    next(action);
    if(action.type === DELETE || ADD || SET){
        saveToStorage(store.getState().users)
    }
};

export default userStorage