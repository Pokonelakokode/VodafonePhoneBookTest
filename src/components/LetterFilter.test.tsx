import * as React from "react";
import * as renderer from "react-test-renderer"
import {createStore, applyMiddleware} from "redux";
import {rootReducer} from "../reducers";
import {hashRouter} from "../middleware/hashRouter";
import userStorage from "../middleware/userStorage";
import { Provider } from "react-redux";
import LetterFilter from "./LetterFilter";

describe("LetterFilter",() => {
    const store = createStore(rootReducer, applyMiddleware(hashRouter, userStorage));
    it('should render snapshot', function () {
        const component = renderer.create(
            <Provider store={store}>
                <LetterFilter/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});