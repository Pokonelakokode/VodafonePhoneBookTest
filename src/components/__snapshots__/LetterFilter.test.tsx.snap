// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Letter should render snapshot 1`] = `
<div
  className="col-12 letters"
>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    C
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    E
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    G
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    K
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    L
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    M
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    N
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    P
  </div>
</div>
`;

exports[`LetterFilter should render snapshot 1`] = `
<div
  className="col-12 letters"
>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    C
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    E
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    G
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    K
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    L
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    M
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    N
  </div>
  <div
    className="letter-selector"
    onClick={[Function]}
  >
    P
  </div>
</div>
`;
